src/main/perl/Component.pm

```diff
@@ -278,6 +277,141 @@ sub _initialize
     return SUCCESS;
 }
 
+=item _redirect
+
+Support component redirection: allow a component to switch
+to a child module based on configuration information.
+
+The component needs a C<REDIRECT> readonly hashref with keys
+
+=over
+
+=item name (mandatory): element name that will be used to lookup
+the child module name in the component configuration. If no value is found,
+and no default is defined, an error is reported and C<_redirect> returns.
+
+=item default (optional): default child module value
+
+=item lower (optional):  when true, no capitalisation of the
+component namespace when generating the child module package name.
+(This does not affect the child module name.)
+
+=back
+
+Caveats
+
+=over
+
+=item NoActionSupported: this is declared by the (parent) component
+and applies for all redirections.
+
+=item C<LC::Exception> context (C<EC>): this is declared
+by the (parent) component, be careful when defining another one in the
+redirections.
+
+=back
+
+Example spma component has 'packager' configuration to switch between
+C<yum>, C<yumng>, C<ips> and C<apt>.
+
+The simplified implementation is as follows:
+
+The C<spma> component package contains
+    package NCM::Component::spma;
+    ...
+    use parent qw(NCM::Component);
+    use Readonly;
+    Readonly our $REDIRECT => {
+        name => 'packager',
+        lower => 1,
+        default => 'yum',
+    }
+
+This means that the module to use is based on the value
+of the C<packager> element in the component tree (based on
+C<prefix()>; in this case C</software/components/spma/packager>).
+In none is defined, the default C<yum> value is used.
+
+The actual module that is loaded from namespace based on the package name,
+with the packagename capitalised.
+In this example, by default this would be in C<NCM::Component::Spma>.
+However, with the C<lower> REDIRECT value true, the capitalisation is
+not performed and the resulting package is C<NCM::Component::spma::yum>.
+
+=cut
+
+sub _redirect
+{
+    my ($self, $method, $config) = @_;
+
+    my $whoami = ref($self);
+
+    my $redirect;
+
+    # TODO: prevent loop?
+    local $@;
+    eval {
+        my $varname = $whoami."::REDIRECT";
+        no strict 'refs';
+        $redirect = ${$varname};
+        use strict 'refs';
+    };
+    if ($@) {
+        $self->verbose("No REDIRECT defined for $whoami: $@");
```

Comment:
When does this get printed? It's only verbose not warn or error so can't be too bad but what are the consequences?

Reply:
it get printed if you do not override `Configure` (so you forgot or you need REDIRECT, in which case you don't need to override Configure). sometimes you think you did set REDIRECT, but it fails, and then it is useful to know what went wrong (well, imho).

Reply:
Feels like an artefact of using an unnatural indirection via a hash, which I think would be avoided by the above?

Reply:
this is some limitation of perl OO. in python this would be

```python
class Magic(object):
    REDIRECT = None
    def moremagic(self):
        is self.REDIRECT is not None:
            return do_something(self.REDIRECT)

class Magic2(Magic):
    REDIRECT = {something else}
```
but you can't do this easily in perl